graphjet-core/src/main/java/com/twitter/graphjet/algorithms/socialproof/EntitySocialProofGenerator.java

```diff
@@ -0,0 +1,178 @@
+/**
+ * Copyright 2017 Twitter. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package com.twitter.graphjet.algorithms.socialproof;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Random;
+
+import com.twitter.graphjet.algorithms.IDMask;
+import com.twitter.graphjet.algorithms.RecommendationAlgorithm;
+import com.twitter.graphjet.algorithms.RecommendationInfo;
+import com.twitter.graphjet.algorithms.RecommendationType;
+import com.twitter.graphjet.algorithms.TweetIDMask;
+import com.twitter.graphjet.bipartite.NodeMetadataLeftIndexedPowerLawMultiSegmentBipartiteGraph;
+import com.twitter.graphjet.bipartite.NodeMetadataMultiSegmentIterator;
+import com.twitter.graphjet.hashing.IntArrayIterator;
+
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectArrayMap;
+import it.unimi.dsi.fastutil.bytes.Byte2ObjectMap;
+import it.unimi.dsi.fastutil.bytes.ByteArraySet;
+import it.unimi.dsi.fastutil.bytes.ByteSet;
+import it.unimi.dsi.fastutil.ints.Int2DoubleMap;
+import it.unimi.dsi.fastutil.ints.Int2DoubleOpenHashMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.ints.IntSet;
+import it.unimi.dsi.fastutil.longs.Long2DoubleMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectArrayMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.LongArraySet;
+import it.unimi.dsi.fastutil.longs.LongSet;
+
+/**
+ * EntitySocialProofGenerator shares similar logic with
+ * {@link com.twitter.graphjet.algorithms.counting.TopSecondDegreeByCount}.
+ * In the request, clients specify a seed user set (left nodes) and an entity set
+ * (right nodes' metadata).
+ * EntitySocialProofGenerator finds the intersection between the seed users' (left node) edges and
+ * the given entity set by traversing each right node's metadata.
+ * Only entities with at least one social proof will be returned to clients.
+ */
+public abstract class EntitySocialProofGenerator implements
```

Comment:
I think the concept of "entity" in GraphJet's namespace is "node metadata". Can we generalize this class to be "NodeMetadataSocialProofGenerator" instead?